src/sudoku.py

```diff
@@ -73,3 +84,24 @@ def __is_valid(self) -> bool:         # проверка на выполнени
                 if not self.__is_valid_cell_sequence(item):
                     return False
         return True
+
+    @property
+    def is_solved(self) -> bool:
+        return all(self.__rows[i][j].is_solved for i in range(9) for j in range(9))
+
+    def __exclude_violating_alternative(self):
+        for i, j in product(range(9), range(9)):
+            if len(self.__rows[i][j].alternatives) == MAX_DEPTH:
+                alternatives = set(self.__rows[i][j].alternatives)
+                while len(alternatives) > 0:
+                    sudoku_copy = copy.deepcopy(self)
+                    sudoku_copy._speculation_depth += 1
+                    sudoku_copy.__rows[i][j].value = alternatives.pop()
```

Comment:
Есть уязвимость, связанная с будущим авто-тестированием и отладкой функционала, в котором есть элемент случайности, т.к. `set.pop()` не предсказуем. Предлагаю заменить на `list.pop()`
```suggestion
                alternatives = sorted(self.__rows[i][j].alternatives)
                while len(alternatives) > 0:
                    sudoku_copy = copy.deepcopy(self)
                    sudoku_copy._speculation_depth += 1
                    sudoku_copy.__rows[i][j].value = alternatives.pop()
```

Reply:
странное дело: в этом случае (alternatives = **sorted**(self.__rows[i][j].alternatives) трудное задание не решается.